File: yasny/pipeline.py
from __future__ import annotations

from .bc import ProgramBC
from .checker import CheckResult, TypeChecker
from .compiler import CompileResult, Compiler
from .lexer import tokenize
from .module_loader import resolve_modules
from .parser import Parser
from .vm import VirtualMachine


def parse_source(source: str, path: str | None = None):
    tokens = tokenize(source, path=path)
    parser = Parser(tokens=tokens, path=path)
    return parser.parse()


def check_program(program, path: str | None = None) -> CheckResult:
    checker = TypeChecker(path=path)
    return checker.check(program)


def compile_program(program, path: str | None = None) -> CompileResult:
    compiler = Compiler(path=path)
    return compiler.compile(program)


def compile_source(source: str, path: str | None = None) -> ProgramBC:
    program = resolve_modules(source, path=path)
    check_program(program, path=path)
    compiled = compile_program(program, path=path)
    return compiled.program


def load_program(source: str, path: str | None = None):
    return resolve_modules(source, path=path)


def run_program(program: ProgramBC, path: str | None = None) -> None:
    vm = VirtualMachine(program=program, path=path)
    vm.run()

File: native/yasn-native/Core/Pipeline.cs
using YasnNative.Bytecode;
using YasnNative.Runtime;

namespace YasnNative.Core;

public static class Pipeline
{
    public static ProgramNode ParseSource(string source, string? path = null)
    {
        var tokens = Lexer.Tokenize(source, path);
        var parser = new Parser(tokens, path);
        return parser.Parse();
    }

    public static ProgramNode LoadProgram(string source, string? path = null)
    {
        var resolver = new ModuleResolver();
        return resolver.ResolveEntry(source, path);
    }

    public static ProgramBC CompileProgram(ProgramNode program, string? path = null)
    {
        TypeChecker.Check(program, path);
        var compiler = new Compiler(path);
        return compiler.Compile(program);
    }

    public static ProgramBC CompileSource(string source, string? path = null)
    {
        var program = LoadProgram(source, path);
        return CompileProgram(program, path);
    }

    public static void RunProgram(ProgramBC program, string? path = null)
    {
        var vm = new VirtualMachine(program, path);
        vm.Run();
    }
}
